tree/ntuple/v7/src/RClusterPool.cxx

```diff
@@ -0,0 +1,329 @@
+{
+   for (const auto &cptr : fPool) {
+      if (cptr && (cptr->GetId() == clusterId))
+         return cptr.get();
+   }
+   return nullptr;
+}
+
+size_t ROOT::Experimental::Detail::RClusterPool::FindFreeSlot() const
+{
+   auto N = fPool.size();
+   for (unsigned i = 0; i < N; ++i) {
+      if (!fPool[i])
+         return i;
+   }
+
+   R__ASSERT(false);
+   return N;
+}
+
+
+namespace {
+
+/// Helper class for the (cluster, column list) pairs that should be loaded in the background
+class RProvides {
+   using DescriptorId_t = ROOT::Experimental::DescriptorId_t;
+   using ColumnSet_t = ROOT::Experimental::Detail::RPageSource::ColumnSet_t;
+
+private:
+   std::map<DescriptorId_t, ColumnSet_t> fMap;
+
+public:
+   void Insert(DescriptorId_t clusterId, const ColumnSet_t &columns)
+   {
+      fMap.emplace(clusterId, columns);
+   }
+
+   bool Contains(DescriptorId_t clusterId) {
+      return fMap.count(clusterId) > 0;
+   }
+
+   void Erase(DescriptorId_t clusterId, const ColumnSet_t &columns)
+   {
+      auto itr = fMap.find(clusterId);
+      if (itr == fMap.end())
+         return;
+      ColumnSet_t d;
+      std::copy_if(itr->second.begin(), itr->second.end(), std::inserter(d, d.end()),
+         [&columns] (DescriptorId_t needle) { return columns.count(needle) == 0; });
+      if (d.empty()) {
+         fMap.erase(itr);
+      } else {
+         itr->second = d;
+      }
+   }
+
+   decltype(fMap)::iterator begin() { return fMap.begin(); }
+   decltype(fMap)::iterator end() { return fMap.end(); }
+};
+
+} // anonymous namespace
+
+ROOT::Experimental::Detail::RCluster *
+ROOT::Experimental::Detail::RClusterPool::GetCluster(
+   DescriptorId_t clusterId, const RPageSource::ColumnSet_t &columns)
+{
+   const auto &desc = fPageSource.GetDescriptor();
+
+   // Determine previous cluster ids that we keep if they happen to be in the pool
+   std::set<DescriptorId_t> keep;
+   auto prev = clusterId;
+   for (unsigned int i = 0; i < fWindowPre; ++i) {
+      prev = desc.FindPrevClusterId(prev);
+      if (prev == kInvalidDescriptorId)
+         break;
+      keep.insert(prev);
+   }
+
+   // Determine following cluster ids and the column ids that we want to make available
+   RProvides provide;
+   provide.Insert(clusterId, columns);
+   auto next = clusterId;
+   for (unsigned int i = 1; i < fWindowPost; ++i) {
+      next = desc.FindNextClusterId(next);
+      if (next == kInvalidDescriptorId)
+         break;
+      provide.Insert(next, columns);
+   }
+
+   // Clear the cache from clusters not the in the look-ahead or the look-back window
+   for (auto &cptr : fPool) {
+      if (!cptr)
+         continue;
+      if (provide.Contains(cptr->GetId()) > 0)
+         continue;
+      if (keep.count(cptr->GetId()) > 0)
+         continue;
+      cptr.reset();
+   }
+
+   // Move clusters that meanwhile arrived into cache pool
+   {
+      std::lock_guard<std::mutex> lockGuardInFlightClusters(fLockWorkQueue);
```

Comment:
This may or may not have a real effect on performance.  This lock seems to be held for a long time in this case here (at least long in  number of lines of code but also 3 loops).

Reply:
I added a clarifying comment:

    // This lock is held during iteration over several data structures: the collection of in-flight clusters,
    // the current pool of cached clusters, and the set of cluster ids to be preloaded.
    // All three collections are expected to be small (certainly < 100, more likely < 10).  All operations
    // are non-blocking and moving around small items (pointers, ids, etc).  Thus the overall locking time should
    // still be reasonably small and the lock is rarely taken (usually once per cluster).